django/contrib/postgres/indexes.py

```diff
@@ -1,47 +1,65 @@
 from django.db.models import Index
+from django.db.utils import NotSupportedError
+from django.utils.functional import cached_property
 
-__all__ = ['BrinIndex', 'GinIndex', 'GistIndex']
+__all__ = ['BrinIndex', 'GinIndex', 'GistIndex', 'HashIndex', 'SpGistIndex']
 
 
-class MaxLengthMixin:
-    # Allow an index name longer than 30 characters since the suffix is 4
-    # characters (usual limit is 3). Since this index can only be used on
-    # PostgreSQL, the 30 character limit for cross-database compatibility isn't
-    # applicable.
-    max_name_length = 31
+class PostgresIndex(Index):
 
+    @cached_property
+    def max_name_length(self):
+        # Allow an index name longer than 30 characters where the suffix is
+        # longer than the usual 3 character limit. The 30 character limit for
+        # cross-database compatibility isn't applicable to indexes only usable
+        # on PostgreSQL.
+        return Index.max_name_length - len(Index.suffix) + len(self.suffix)
```

Comment:
Can we restrict the suffixes of these indexes to 3 characters as well (to be consistent with other non-postgres code)?
What benefits are we gaining from increasing the limit of the suffix part of the index name - and why _just_ the suffix part then, why not allow longer names as well if we're making it postgres-specific.

Although I would try to keep the rules as consistent as possible and simply give 3 chars suffixes to these indexes as well; Or maybe I am missing something here?

Reply:
I think not for a few reasons:

- The ship has sailed - we are already using longer suffixes for PostgreSQL, e.g. `brin` and `like`.
- The name truncating code is part of the main base `Index`. It would need separate logic for PostgreSQL which is not impossible, but beyond the scope of these tickets.
- Suffixes such as `brn`, `hsh`, `gst`, `spg`, `lik` would just be unclear for no good reason other than unrelated backends having a name length restriction.

Reply:
A reason for the longer suffixes was to simplify inspectdb. Using different names would require mapping the names that PostgreSQL returns to whatever names we picked.

Reply:
I see. Thanks @pope1ni and @timgraham